pkg/runtime/runtime.go

```
@@ -1179,7 +1179,13 @@ func (a *DaprRuntime) publishMessageGRPC(msg *pubsub.NewMessage) error {
 	}
 
 	if data, ok := cloudEvent[pubsub.DataField]; ok && data != nil {
-		envelope.Data = []byte(cloudEvent[pubsub.DataField].(string))
+		envelope.Data = nil
+
+		if contenttype.IsStringContentType(envelope.DataContentType) {
+			envelope.Data = []byte(data.(string))
+		} else if contenttype.IsJSONContentType(envelope.DataContentType) {
+			envelope.Data, _ = a.json.Marshal(data)
```

@yaron2 I think the error should not be swallowed here.